### PR TITLE
US016 edit capability

### PR DESCRIPTION
**Pull request by caitlinbulloch**
**Edit a Capability Name**

- created capability controller
- created capability service
- created capability model
- created capability list, edit, and confirm html pages. 

Screenshot 2024-01-10 at 16 24 31

Screenshot 2024-01-10 at 16 25 05

Screenshot 2024-01-10 at 16 25 13

### CODE REVIEW

File: app.ts [hunk 1]
```diff
@@ -29,8 +31,12 @@ app.use(express.json());
 
 app.use(express.urlencoded({ extended: false }));
 
+app.use(session({ secret: "NOT HARDCODED SECRET", cookie: { maxAge: 60000 }}));
```
Comment (eoinmcbennett) [inline app.ts]: Watch for this in a merge conflict, use the config value for this.

File: controller/capabilityController.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        }
+        res.render("edit-capability", { capability: data } );
+    });
+
+    app.get("/edit-capability", async (req: Request, res: Response) => {
+        if (!req.session.capability) {
+            req.session.capability = {}
+        }
+        res.render("edit-capability")
+    })
```
Comment (eoinmcbennett) [inline controller/capabilityController.ts]: Does this ever get used?

File: service/capabilityService.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import { Responsibility } from "../model/responsibility";
+import { Capability } from "../model/capability";
+
+const axios = require("axios");
+const capabilityValidator = require("../Validator/capabilityValidator");
+
+
+module.exports.getAllCapabilities = async function (): Promise<Capability> {
+    try {
+        const response = await axios.get("http://localhost:8080/api/capability");
```
Comment (eoinmcbennett) [inline service/capabilityService.ts]: This url will need to be changed to use the env API_URL or it wont work with AWS backend.

File: service/capabilityService.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        return response.data;
+    } catch (e) {
+        throw new Error("Could not get capabilities")
+    }
+};
+
+module.exports.getCapabilityById = async function (id: number): Promise<Capability> {
+    try {
+        const response = await axios.get("http://localhost:8080/api/capability/" + id);
```
Comment (eoinmcbennett) [inline service/capabilityService.ts]: This url will need to be changed to use the env API_URL or it wont work with AWS backend.

File: service/capabilityService.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    }
+};
+
+module.exports.updateCapability = async function (capability: Capability ): Promise<number> {
+    const error: string = capabilityValidator.validateCapability(capability)
+    if(error) {
+        throw new Error(error)
+    }
+    try {
+        const response = await axios.put("http://localhost:8080/api/capability/" + capability.capabilityID , capability)
```
Comment (eoinmcbennett) [inline service/capabilityService.ts]: This url will need to be changed to use the env API_URL or it wont work with AWS backend.

**Review by eoinmcbennett — CHANGES_REQUESTED** [review]
Looks great, can you just take a look at some of the things I flagged.

**Review by eoinmcbennett — APPROVED** [review]

File: app.ts [hunk 6]
```diff
@@ -14,6 +14,12 @@ const app = express();
 const session = require("express-session");
 const bodyParser = require("body-parser");
 
+declare module "express-session" {
+    interface SessionData {
+        isLoggedIn: boolean;
+    }
+}
+
```
Comment (eoinmcbennett) [inline app.ts]: Having multiple of these can break things, consolidate them into one.

File: app.ts [hunk 7]
```diff
@@ -54,6 +60,7 @@ declare module "express-session" {
     interface SessionData {
         job: Job
         capability: Capability
+        token: string;
     }
 }
```
Comment (eoinmcbennett) [inline app.ts]: Having multiple of these can break things, consolidate them into one.

**Review by eoinmcbennett — CHANGES_REQUESTED** [review]

File: app.ts
Comment (eoinmcbennett) [inline app.ts:1]: Dont do app.use(session) twice, could break things.

**Review by eoinmcbennett — CHANGES_REQUESTED** [review]

**Review by eoinmcbennett — APPROVED** [review]

**Review by Cormac-F — APPROVED** [review]

**Pull request merged**